src/foam/nanos/crunch/ui/CapabilityView.js

```diff
@@ -0,0 +1,105 @@
+foam.CLASS({
+  package: 'foam.nanos.crunch.ui',
+  name: 'MyView',
+  extends: 'foam.u2.View',
+  documentation: 'All purpose capability view that takes in an array of capaiblity ids and display these capabilities',
+
+  imports: [
+    'crunchController',
+    'notify'
+  ],
+
+  requires: [
+    'foam.u2.detail.SectionView'
+  ],
+
+  properties: [
+    {
+      class: 'StringArray',
+      name: 'capabilities'
+    },
+    {
+      name: 'wizardlets',
+      documentation: 'wizardlets for the capability'
+    },
+    {
+      name: 'wizardletSectionsList',
+      documentation: `
+        sections for wizardlets
+        wizardletSectionsList[i] stores sections for wizardlets[i]
+      `
+    }
+  ],
+
+  methods: [
+    async function initE() {
+      const self = this;
+
+      this.start().addClass(this.myClass())
+        .add(this.slot(function(capabilities) {
+          return this.E().forEach(capabilities, async function(cap) {
+            // get wizardlets for the capability and all of its dependent capabilities
+            self.wizardlets = (await self.crunchController.getCapsAndWizardlets(cap)).wizCaps;
+
+            // add listeners on wizardlets
+            for ( const wizardlet of self.wizardlets ) {
+              self.addListeners(wizardlet.data);
+            }
+
+            // get all the sections associated with the wizardlets
+            self.wizardletSectionsList = self.crunchController.generateSections(self.wizardlets);
+            
+            this.add(self.slot(function(wizardletSectionsList) {
+              return this.E().forEach(wizardletSectionsList, function(sections, index) {
+                sections.map(section => (
+                  this.tag(self.SectionView, {
+                    section,
+                    data: self.wizardlets[index].data,
+                    showTitle: false
+                  })
+                ));
+              });
+            }));
+          });
+        }))
+      .end();
+    },
+
+    function addListeners(obj) {
+      if ( ! this.isFObject(obj) ) return;
+
+      obj.sub(this.updateWizardlet);
+
+      // add listeners on inner fobjects
+      for ( const value of Object.values(obj.instance_) ) {
+        if ( this.isFObject(value) ) {
+          this.addListeners(value);
+        }
+      }
+      return;
+    },
+
+    function isFObject(obj) {
+      if ( typeof obj === 'object' && obj.instance_ ) {
+        return true;
+      }
+    }
+  ],
+
+  listeners: [
+    {
+      name: 'updateWizardlet',
+      code: function() {
+        // TODO: can we only update the wizardlet that has new data instead of updating all the wizardlets?
```

Comment:
think what you have works. The listener listens to changes. When the data changes we can assume new data.

Reply:
What I am doing here is if one wizardlet is updated, we save all the wizardlets which seemed inefficient to me. I was wondering if there is a way to know on which wizardlet the listener was triggered so that I can only save that wizardlet.

Reply:
First good catch, sorry I missed that.

So take a look at **com.google.foam.demos.sweeper.Board**, it will give an example of passing in a value to a listener. I would suggest adjusting code to pass in wizardlet into the listener then just saving that wizardlet.

Also think adding an init method and moving the adding the listeners to that. initE can get called multiple times.